torch/testing/asserts.py

```diff
@@ -0,0 +1,186 @@
+import warnings
+from typing import Callable, Optional, Tuple
+
+import torch
+from torch.testing._internal.common_utils import get_comparison_dtype as _get_comparison_dtype, TestCase as _TestCase
+
+__all__ = ["assert_tensors_equal", "assert_tensors_allclose"]
+
+try:
+    import pytest
+
+    UsageError = pytest.UsageError
+except ImportError:
+
+    class UsageError(Exception):
+        pass
+
+
+class UsageWarning(Warning):
+    pass
+
+
+def _get_default_rtol_and_atol(tensor1: torch.Tensor, tensor2: torch.Tensor) -> Tuple[float, float]:
+    return _TestCase()._getDefaultRtolAndAtol(tensor1.dtype, tensor2.dtype)
+
+
+def _hide_internal_traceback(fn):
+    @functools.wraps(fn)
+    def wrapper(*args, **kwargs):
+        try:
+            return fn(*args, **kwargs)
+        except AssertionError:
+            for frame_info in inspect.trace():
+                frame_info.frame.f_locals.setdefault("__tracebackhide__", True)
+            raise
+
+    return wrapper
+
+
+# These attributes will be checked by default in every call to assert_tensors_(equal|allclose) and
+# assert_(equal|allclose) if the inputs are tensors.
+DEFAULT_CHECKED_ATTRIBUTES = (
+    "shape",
+    "dtype",
+)
+
+
+def _compare_tensors_meta(tensor1: torch.Tensor, tensor2: torch.Tensor, **attributes: bool) -> None:
+    if not (isinstance(tensor1, torch.Tensor) and isinstance(tensor2, torch.Tensor)):
+        raise AssertionError(f"Both inputs have to be tensors, but got {type(tensor1)} and {type(tensor2)} instead.")
+
+    for attr in DEFAULT_CHECKED_ATTRIBUTES:
+        attributes.setdefault(attr, True)
+
+    for attr, check in attributes.items():
+        if not check:
+            continue
+
+        try:
+            val1 = getattr(tensor1, attr)
+            val2 = getattr(tensor2, attr)
+            if val1 != val2:
+                raise AssertionError(f"The values for attribute '{attr}' do not match: {val1} != {val2}.")
+        except AttributeError as error:
+            raise UsageError(f"At least one of the inputs does not have an attribute '{attr}'.") from error
+
+
+def _compare_tensors(
+    tensor1: torch.Tensor,
+    tensor2: torch.Tensor,
+    asserter: Callable[[torch.Tensor, torch.Tensor], None],
+    **strict_attributes: bool,
+) -> None:
+    if any(tensor.dtype in (torch.complex32, torch.complex64, torch.complex128) for tensor in (tensor1, tensor2)):
+        raise UsageError("Comparison for complex tensors is not supported yet.")
+    if any(tensor.is_quantized for tensor in (tensor1, tensor2)):
+        raise UsageError("Comparison for quantized tensors is not supported yet.")
+    if any(tensor.is_sparse for tensor in (tensor1, tensor2)):
+        raise UsageError("Comparison for sparse tensors is not supported yet.")
+
+    _compare_tensors_meta(tensor1, tensor2, **strict_attributes)
+
+    if tensor1.device != tensor2.device:
+        tensor1 = tensor1.cpu()
+        tensor2 = tensor2.cpu()
+
+    if tensor1.dtype != tensor2.dtype:
```

Comment:
Why is this necessary? Does torch.eq not support type promotion?

Reply:
I took this from our current implementation

https://github.com/pytorch/pytorch/blob/1772e26f6380d189ba4e03da492ff18eaa139563/torch/testing/_internal/common_utils.py#L1050-L1054

Plus, we also need `torch.allclose` to support it. I'll investigate and remove if it unnecessary.

Reply:
Just checked, `torch.allclose` does *not* support type promotion.

```python
import torch

a = torch.empty((1, 2, 3), dtype=torch.uint8)
b = a.clone().float()

assert torch.all(torch.eq(a, b))
assert torch.allclose(a, b)
```

```
RuntimeError: Byte did not match Float
```

Given that `torch.eq` supports it, shouldn't `torch.allclose` also support it? @mruberry do you know who could have insights into this?

Reply:
Thank you for checking. Providing it ourselves sounds like the right approach, then. Would you make a note?

I'm actually about to publish an RFC for which ops do/don't support type promotion. My thinking was that unary ufuncs, binary ufuncs, and reductions should and other ops shouldn't. That would suggest that torch.allclose wouldn't support type promotion.

Reply:
In that case would it make sense to have a `torch.close` function that doesn't reduce the input and thus supports type promotion? For equality checking we are simply using `torch.all(torch.eq(a, b))`

Reply:
That's why we have [isclose()](https://pytorch.org/docs/master/generated/torch.isclose.html?highlight=isclose#torch.isclose), like [NumPy does](https://numpy.org/doc/stable/reference/generated/numpy.isclose.html?highlight=isclose#numpy.isclose).

Note, however, that isclose() does not handle complex comparisons like our current test suite does:

https://github.com/pytorch/pytorch/blob/bb21aea37add0400eaa4ea8317656b7469b38a94/aten/src/ATen/native/TensorCompare.cpp#L24

Reply:
That is news to me. Thanks for the heads-up.